<a href="https://colab.research.google.com/github/IKSectors/Repo/blob/main/Observations.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **Exploratory Data Analysis (EDA)**

An ML model usually starts with a data preparation step in which the loading of the key libraries is considered. This phase is known as exploratory data analysis, or EDA. Numpy, Pandas, and Matplotlib are present. Pyplot is the foundation of the model design, which allows the designer to easily analyse the data.

We will be performing EDA from the dataset to change the data into usable form (also known as data preparation) for analytical operations.


### **Libraries & Packages**
Let's start by installing the libraries and packages required for the ML model.

In [1]:
import pandas as pd
import numpy as np
import tensorflow as tf
import matplotlib.pyplot as plt
from sklearn.manifold import TSNE
from sklearn.decomposition import PCA

pd.set_option('display.max_columns', None)

In [10]:
!pip install scikit-learn
!pip install seaborn
!pip install nltk
!pip install beautifulsoup4
!pip install -U ydata_profiling

Data has to be structured in most cases to make data analysis possible. The data we are using was imported from Kaggle's website and has to be read into a dataframe to enable analysis and subsequent hypothesis to be made. The above python libraries have to be imported to be able to convert the dataset into a pandas dataframe.
But before we get to that, let's mount locate and make available the imported Kaggle file.


Next, we mount the goodle drive where our files are located...

In [11]:
from google.colab import drive
drive.mount('/content/drive')
%matplotlib inline

Mounted at /content/drive


Drive has been mounted successfully at /content/drive location instance... next is to locate and run the data file from Kaggle...

In [12]:
data = pd.read_excel("/content/drive/MyDrive/UWL ML Project 2023/Resources/Kaggle_ICU_Dataset.xlsx")
data

PATIENT_VISIT_IDENTIFIER  AGE_ABOVE65 AGE_PERCENTIL  GENDER  \
0                            0            1          60th       0   
1                            0            1          60th       0   
2                            0            1          60th       0   
3                            0            1          60th       0   
4                            0            1          60th       0   
...                        ...          ...           ...     ...   
1920                       384            0          50th       1   
1921                       384            0          50th       1   
1922                       384            0          50th       1   
1923                       384            0          50th       1   
1924                       384            0          50th       1   

      DISEASE GROUPING 1  DISEASE GROUPING 2  DISEASE GROUPING 3  \
0                    0.0                 0.0                 0.0   
1                    0.0                 0.0                 0.0   
2                    0.0                 0.0                 0.0   
3                    0.0                 0.0                 0.0   
4                    0.0                 0.0                 0.0   
...                  ...                 ...                 ...   
1920                 0.0                 0.0                 0.0   
1921                 0.0                 0.0                 0.0   
1922                 0.0                 0.0                 0.0   
1923                 0.0                 0.0                 0.0   
1924                 0.0                 0.0                 1.0   

      DISEASE GROUPING 4  DISEASE GROUPING 5  DISEASE GROUPING 6  HTN  \
0                    0.0                 1.0                 1.0  0.0   
1                    0.0                 1.0                 1.0  0.0   
2                    0.0                 1.0                 1.0  0.0   
3                    0.0                 1.0                 1.0  0.0   
4                    0.0                 1.0                 1.0  0.0   
...                  ...                 ...                 ...  ...   
1920                 0.0                 0.0                 0.0  0.0   
1921                 0.0                 0.0                 0.0  0.0   
1922                 0.0                 0.0                 0.0  0.0   
1923                 0.0                 0.0                 0.0  0.0   
1924                 0.0                 0.0                 0.0  0.0   

      IMMUNOCOMPROMISED  OTHER  ALBUMIN_MEDIAN  ALBUMIN_MEAN  ALBUMIN_MIN  \
0                   0.0    1.0             NaN           NaN          NaN   
1                   0.0    1.0             NaN           NaN          NaN   
2                   0.0    1.0        0.605263      0.605263     0.605263   
3                   0.0    1.0             NaN           NaN          NaN   
4                   0.0    1.0        0.000000      0.000000     0.000000   
...                 ...    ...             ...           ...          ...   
1920                0.0    1.0             NaN           NaN          NaN   
1921                0.0    1.0        0.605263      0.605263     0.605263   
1922                0.0    1.0             NaN           NaN          NaN   
1923                0.0    1.0             NaN           NaN          NaN   
1924                0.0    1.0        0.605263      0.605263     0.605263   

      ALBUMIN_MAX  ALBUMIN_DIFF  BE_ARTERIAL_MEDIAN  BE_ARTERIAL_MEAN  \
0             NaN           NaN                 NaN               NaN   
1             NaN           NaN                 NaN               NaN   
2        0.605263          -1.0           -1.000000         -1.000000   
3             NaN           NaN                 NaN               NaN   
4        0.000000          -1.0           -0.871658         -0.871658   
...           ...           ...                 ...               ...   
1920          NaN           NaN                

The Kaggle file has been read successfully and it produced 1925 rows x 231 columns. We can get a more detailed information from the info() function below...

In [13]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1925 entries, 0 to 1924
Columns: 231 entries, PATIENT_VISIT_IDENTIFIER to ICU
dtypes: float64(225), int64(4), object(2)
memory usage: 3.4+ MB


The dataset produced a RangeIndex, otherwise called rows of 1925 entries, which ranges from 0 to 1924.
It comprises of 231 columns which can further be broken down into data types (dtypes) of 225 floats, 4 integers and 2 objects.


The relationships between the variables will be observed to enable the analysis of the data to be done. The first simple command to check for this will be the describe() command.



In [14]:
data.describe()

PATIENT_VISIT_IDENTIFIER  AGE_ABOVE65       GENDER  DISEASE GROUPING 1  \
count               1925.000000  1925.000000  1925.000000         1920.000000   
mean                 192.000000     0.467532     0.368831            0.108333   
std                  111.168431     0.499074     0.482613            0.310882   
min                    0.000000     0.000000     0.000000            0.000000   
25%                   96.000000     0.000000     0.000000            0.000000   
50%                  192.000000     0.000000     0.000000            0.000000   
75%                  288.000000     1.000000     1.000000            0.000000   
max                  384.000000     1.000000     1.000000            1.000000   

       DISEASE GROUPING 2  DISEASE GROUPING 3  DISEASE GROUPING 4  \
count         1920.000000         1920.000000         1920.000000   
mean             0.028125            0.097917            0.019792   
std              0.165373            0.297279            0.139320   
min              0.000000            0.000000            0.000000   
25%              0.000000            0.000000            0.000000   
50%              0.000000            0.000000            0.000000   
75%              0.000000            0.000000            0.000000   
max              1.000000            1.000000            1.000000   

       DISEASE GROUPING 5  DISEASE GROUPING 6          HTN  IMMUNOCOMPROMISED  \
count         1920.000000         1920.000000  1920.000000        1920.000000   
mean             0.128125            0.046875     0.213021           0.158333   
std              0.334316            0.211426     0.409549           0.365148   
min              0.000000            0.000000     0.000000           0.000000   
25%              0.000000            0.000000     0.000000           0.000000   
50%              0.000000            0.000000     0.000000           0.000000   
75%              0.000000            0.000000     0.000000           0.000000   
max              1.000000            1.000000     1.000000           1.000000   

             OTHER  ALBUMIN_MEDIAN  ALBUMIN_MEAN  ALBUMIN_MIN  ALBUMIN_MAX  \
count  1920.000000      821.000000    821.000000   821.000000   821.000000   
mean      0.809896        0.528527      0.528527     0.528527     0.528527   
std       0.392485        0.224100      0.224100     0.224100     0.224100   
min       0.000000       -1.000000     -1.000000    -1.000000    -1.000000   
25%       1.000000        0.605263      0.605263     0.605263     0.605263   
50%       1.000000        0.605263      0.605263     0.605263     0.605263   
75%       1.000000        0.605263      0.605263     0.605263     0.605263   
max       1.000000        1.000000      1.000000     1.000000     1.000000   

       ALBUMIN_DIFF  BE_ARTERIAL_MEDIAN  BE_ARTERIAL_MEAN  BE_ARTERIAL_MIN  \
count         821.0          821.000000        821.000000       821.000000   
mean           -1.0           -0.963433         -0.963433        -0.963433   
std             0.0            0.160870          0.160870         0.160870   
min            -1.0           -1.000000         -1.000000        -1.000000   
25%            -1.0           -1.000000         -1.000000        -1.000000   
50%            -1.0           -1.000000         -1.000000        -1.000000   
75%            -1.0           -1.000000         -1.000000        -1.000000   
max            -1.0            1.000000          1.000000         1.000000   

       BE_ARTERIAL_MAX  BE_ARTERIAL_DIFF  BE_VENOUS_MEDIAN  BE_VENOUS_MEAN  \
count       821.000000             821.0        821.000000      821.000000   
mean         -0.963433              -1.0         -0.931121       -0.931121   
std           0.160870               0.0          0.169509        0.169509   
min          -1.000000              -1.0         -1.000000       -1.000000   
25%          -1.000000              -1.0         -1.000000       -1.000000   
50%          -1.000000              -1.0         -1


The above cell has ran the data.describe() command which produced the mean, mode, median, and std values of the variables contained in the dataset.

# Categorical Variables
Categorical variables are variables that can take on values from a limited and usually fixed set of categories. For example, a "Color" variable could have categories like "Red," "Green," and "Blue."

When dealing with categorical variables in a dataset, they need to be converted into a numerical format before feeding them into machine learning algorithms, as most algorithms work with numerical data. One common method for handling categorical variables is one-hot encoding.

The next steps are to analyse the columns to see how they can be categorized.

In [15]:
data.columns

Index(['PATIENT_VISIT_IDENTIFIER', 'AGE_ABOVE65', 'AGE_PERCENTIL', 'GENDER',
       'DISEASE GROUPING 1', 'DISEASE GROUPING 2', 'DISEASE GROUPING 3',
       'DISEASE GROUPING 4', 'DISEASE GROUPING 5', 'DISEASE GROUPING 6',
       ...
       'TEMPERATURE_DIFF', 'OXYGEN_SATURATION_DIFF',
       'BLOODPRESSURE_DIASTOLIC_DIFF_REL', 'BLOODPRESSURE_SISTOLIC_DIFF_REL',
       'HEART_RATE_DIFF_REL', 'RESPIRATORY_RATE_DIFF_REL',
       'TEMPERATURE_DIFF_REL', 'OXYGEN_SATURATION_DIFF_REL', 'WINDOW', 'ICU'],
      dtype='object', length=231)

Let's also see the datatypes in a comprehensive view with the dtypes() command

In [16]:
print(data.dtypes)

PATIENT_VISIT_IDENTIFIER        int64
AGE_ABOVE65                     int64
AGE_PERCENTIL                  object
GENDER                          int64
DISEASE GROUPING 1            float64
                               ...   
RESPIRATORY_RATE_DIFF_REL     float64
TEMPERATURE_DIFF_REL          float64
OXYGEN_SATURATION_DIFF_REL    float64
WINDOW                         object
ICU                             int64
Length: 231, dtype: object


Next, let's select the OBJECT datatypes

In [17]:
data.select_dtypes(object)

AGE_PERCENTIL    WINDOW
0             60th       0-2
1             60th       2-4
2             60th       4-6
3             60th      6-12
4             60th  ABOVE_12
...            ...       ...
1920          50th       0-2
1921          50th       2-4
1922          50th       4-6
1923          50th      6-12
1924          50th  ABOVE_12

[1925 rows x 2 columns]

# **Let's briefly describe the types of columns you have:**

**Numerical Features (int64 and float64)**:

PATIENT_VISIT_IDENTIFIER: An integer identifier for each patient visit.
AGE_ABOVE65: A binary indicator (0 or 1) indicating whether the patient is above 65 years old.
GENDER: A binary indicator (0 or 1) for gender, where 0 may represent one gender, and 1 may represent another.
DISEASE GROUPING 1, ...: Numerical features related to different disease groupings.
**Categorical Features (object)**:
**AGE_PERCENTIL**: A categorical feature representing age percentiles.

Binary Feature (int64):
ICU: A binary indicator (0 or 1) indicating whether the patient was admitted to the ICU.
**Temporal Feature (object)**:
**WINDOW**: A categorical feature indicating the time window of the observation.
Before using this data in a machine learning model, you might want to preprocess it. Specifically, you may need to handle missing values, encode categorical variables (like AGE_PERCENTIL), and decide whether to use one-hot encoding or another encoding method.

**Here's a general outline of steps you might consider**:

Handle Missing Values:
Check for missing values in your dataset and decide on an appropriate strategy, such as removing rows or columns with missing values, imputing missing values, or using more advanced imputation methods.

Encode Categorical Variables:
Encode categorical variables like AGE_PERCENTIL. You can use one-hot encoding, label encoding, or another suitable method based on the nature of the data and your modeling requirements.

Feature Scaling:
Depending on the machine learning algorithms you plan to use, you might need to scale numerical features for better model performance. Common scaling methods include Min-Max scaling or standardization.

Split Data:
If you're building a predictive model, split your dataset into training and testing sets to evaluate the model's performance on unseen data.

Remember that the specific preprocessing steps may vary depending on the nature of your data and the goals of your analysis or model. Always explore and understand your data before applying any preprocessing techniques.

**Bias and Fairness:**

One-Hot Encoding: This method creates a binary column for each category, potentially leading to a large number of binary features. If certain categories are underrepresented in the dataset, it might result in imbalanced features, potentially leading to biased model outcomes.
Binary Encoding: In contrast, binary encoding may result in fewer features and can be more memory-efficient. However, if certain categories have more weight due to their binary representation, it may introduce bias.
Transparency and Interpretability:

**One-Hot Encoding**:

One-hot encoding provides a clear and interpretable representation of categorical variables. Each category is explicitly represented by its own binary column, which can be beneficial for understanding model predictions.
Binary Encoding: Binary encoding might make the model less transparent since the relationship between the binary representation and the original categories is not as straightforward.

**Privacy Concerns:**

One-Hot Encoding: Creating a binary column for each category might reveal more information about individual instances in the dataset, potentially raising privacy concerns, especially if the dataset contains sensitive information.
Binary Encoding: Binary encoding can mitigate privacy concerns to some extent, as it represents categories with binary codes that might be less revealing about individual instances.
Algorithmic Impact:

One-Hot Encoding: Some machine learning algorithms may be sensitive to the high dimensionality introduced by one-hot encoding, potentially impacting model training and performance.
Binary Encoding: Binary encoding can result in a more compact representation, which might be preferred for certain algorithms, especially those that struggle with high-dimensional data.
Data Representation Impact:

**One-Hot Encoding**: The explicit representation of each category in one-hot encoding can impact how the data is perceived, especially if the categories have specific cultural or social connotations.
Binary Encoding: The compressed representation in binary encoding may reduce the interpretability of the original categories, potentially leading to a loss of nuance.
When choosing between one-hot encoding and binary encoding, it's important to consider the specific ethical implications within the context of your data, the problem domain, and the potential impact on individuals or groups represented in the dataset

# **Data Preparation**

The following steps will be used to transform the raw data into a format that may be used:

To transform, I will use binary hotcoding rather than float columns. In Data preparation step where I plan to use binary encoding (also known as one-hot encoding) for transforming categorical variables into a format suitable for machine learning models. Binary encoding is a method used to represent categorical variables with multiple categories as binary vectors.

Here's a general outline of the steps involved in binary encoding:

Identify Categorical Variables: Identify which columns in your raw data are categorical variables. These are typically variables with discrete values, such as colors, categories, or labels.

Label Encoding (if needed): Before one-hot encoding, you may need to perform label encoding for ordinal categorical variables, where the order matters. Label encoding assigns a unique integer to each category.

One-Hot Encoding: For each categorical variable, apply one-hot encoding. This involves creating binary columns for each category and representing the presence of that category with a 1 and its absence with a 0.

Drop Original Categorical Columns: After one-hot encoding, you can drop the original categorical columns since they are now represented by the new binary columns.

Data Scaling (if needed): Depending on your machine learning algorithm, you might need to scale your features. For some algorithms, especially those based on distances or gradients, it's essential to have features on a similar scale.

**Handle Missing Values (if needed): Check and handle any missing values in your dataset. Depending on the amount of missing data, you might choose to impute values or drop the corresponding rows or columns.

**If the patient has been admitted to the ICU in any of the upcoming windows, mark Window 0–2 as 1.**